src/Processors/QueryPlan/Optimizations/reduceAggregationKeysSize.cpp

```diff
@@ -0,0 +1,176 @@
+
+            changed_data_types.push_back(key_column_type);
+            ColumnWithTypeAndName const_column;
+            if (hint.lower_boundary->getTypeName() == "Int64")
+                const_column.type = std::make_shared<DataTypeInt64>();
+            else
+                const_column.type = std::make_shared<DataTypeUInt64>();
+            const_column.column = const_column.type->createColumnConst(1, hint.lower_boundary.value());
+            const_column.name = "__minus_value_" + old_key;
+            const auto * hint_node = &dag->addColumn(const_column);
+
+            ActionsDAG::NodeRawConstPtrs children = {key_column_node, hint_node};
+            auto minus_function = FunctionFactory::instance().get("minus", Context::getGlobalContextInstance());
+            const auto & added_function = dag->addFunction(minus_function, children, "__hinted_key_uncasted_" + old_key);
+
+            DataTypePtr result_type;
+            if (hint.isRangeLengthLessOrEqualThan(255))
+                result_type = std::make_shared<DataTypeUInt8>();
+            else
+                result_type = std::make_shared<DataTypeUInt16>();
+
+            dag->addOrReplaceInOutputs(dag->addCast(added_function, result_type, new_key));
+        }
+        else
+        {
+            new_aggregating_keys.push_back(old_key);
+        }
+    }
+
+    return dag;
+}
+
+std::shared_ptr<ActionsDAG> buildSuccessorActionsDag(const Block & header, const Names & changed_keys, const DataTypes& changed_data_types, const DataHints & hints)
+{
+    auto dag = std::make_shared<ActionsDAG>(header.getColumnsWithTypeAndName());
+
+    for (size_t i = 0; i < changed_keys.size(); ++i)
+    {
+        const auto & changed_key = changed_keys[i];
+        const auto & changed_data_type = changed_data_types[i];
+
+        const auto & hint = hints.at(changed_key);
+
+        const auto key_to_remove = "__hinted_key_" + changed_key;
+
+        const auto * node_to_remove = &dag->findInOutputs(key_to_remove);
+        ColumnWithTypeAndName const_column;
+        if (hint.lower_boundary->getTypeName() == "Int64")
+            const_column.type = std::make_shared<DataTypeInt64>();
+        else
+            const_column.type = std::make_shared<DataTypeUInt64>();
+        const_column.column = const_column.type->createColumnConst(1, hint.lower_boundary.value());
+        const_column.name = "__plus_value_" + changed_key;
+        const auto * hint_node = &dag->addColumn(const_column);
+
+        ActionsDAG::NodeRawConstPtrs children = {node_to_remove, hint_node};
+        auto plus_function = FunctionFactory::instance().get("plus", Context::getGlobalContextInstance());
+        dag->addOrReplaceInOutputs(dag->addCast(dag->addFunction(plus_function, children, "__uncasted_" + changed_key), changed_data_type, changed_key));
+        dag->removeUnusedResult(key_to_remove);
+    }
+
+    return dag;
+}
+
+}
+
+size_t tryReduceAggregationKeysSize(QueryPlan::Node * node, QueryPlan::Nodes & nodes)
```

Comment:
pls create a comment describing the transformation. there are tools like https://asciiflow.com/ that allow to create a visual scheme using ascii charset.
in general we like to write comments.